client/lib/plugins/sanitize-section-content.js

```diff
@@ -0,0 +1,190 @@
+/** @format */
+/**
+ * External dependencies
+ */
+import validUrl from 'valid-url';
+
+const isAllowedTag = tagName => {
+	switch ( tagName ) {
+		case '#text':
+		case 'a':
+		case 'b':
+		case 'blockquote':
+		case 'code':
+		case 'div':
+		case 'em':
+		case 'h1':
+		case 'h2':
+		case 'h3':
+		case 'h4':
+		case 'h5':
+		case 'h6':
+		case 'i':
+		case 'img':
+		case 'li':
+		case 'ol':
+		case 'p':
+		case 'span':
+		case 'strong':
+		case 'ul':
+			return true;
+		default:
+			return false;
+	}
+};
+
+const isAllowedAttr = ( tagName, attrName ) => {
+	if ( 'a' === tagName && 'href' === attrName ) {
+		return true;
+	}
+
+	if ( 'iframe' === tagName ) {
+		switch ( attrName ) {
+			case 'class':
+			case 'type':
+			case 'height':
+			case 'width':
+			case 'src':
+				return true;
+			default:
+				return false;
+		}
+	}
+
+	if ( 'img' === tagName && ( 'alt' === attrName || 'src' === attrName ) ) {
+		return true;
+	}
+
+	return false;
+};
+
+const isValidYoutubeEmbed = node => {
+	if ( node.nodeName.toLowerCase() !== 'iframe' ) {
+		return false;
+	}
+
+	if ( node.getAttribute( 'class' ) !== 'youtube-player' ) {
+		return false;
+	}
+
+	if ( node.getAttribute( 'type' ) !== 'text/html' ) {
+		return false;
+	}
+
+	const link = document.createElement( 'a' );
+	link.href = node.getAttribute( 'src' );
+
+	return (
+		validUrl.isWebUri( node.getAttribute( 'src' ) ) &&
+		( link.hostname === 'youtube.com' || link.hostname === 'www.youtube.com' )
+	);
+};
+
+/**
+ * Sanitizes input HTML for security and styling
+ *
+ * @param {String} content unverified HTML
+ * @returns {string} sanitized HTML
+ */
+export const sanitizeSectionContent = content => {
+	const parser = new DOMParser();
+	const doc = parser.parseFromString( content, 'text/html' );
+
+	// this will let us visit every single DOM node programmatically
+	const walker = doc.createTreeWalker( doc.body );
+
+	// we don't want to remove nodes while walking the tree
+	// or we'll invite data-race bugs. instead, we'll track
+	// which ones we want to remove then drop them at the end
+	const removeList = [];
+
+	// walk over every DOM node
+	while ( walker.nextNode() ) {
+		const node = walker.currentNode;
+		const tagName = node.nodeName.toLowerCase();
+
+		if ( ! isAllowedTag( tagName ) && ! isValidYoutubeEmbed( node ) ) {
+			removeList.push( node );
+			continue;
+		}
+
+		// strip out anything not explicitly allowed
+		// in the attributes. we want to eliminate
+		// potential cross-site scripting attacks _and_
+		// prevent custom styles from interfering with
+		// our page's own rendering
+
+		const attrRemoveList = [];
+
+		// these aren't Arrays, so iteration is odd
+		for ( let i = 0; node.attributes && i < node.attributes.length; i++ ) {
```

Comment:
If you want to iterate functionally here, then lodash is your friend :wink:
```js
_.forEach( node.attributes, ( { name, value } ) => ... );
```
Or create the `attrRemoveList` by filtering:
```js
const attrRemoveList = _.filter( node.attributes, ... );
attrRemoveList.forEach( ( { name } ) => node.remove(...) );
```

Reply:
lodash'd in 6e9d080df

Reply:
I chose not to use the filter because of how the flow of the function goes. not sure if this is the right decision, but it's odd to me because we're not just walking over the filtered or non-filtered items, we're walking through the tree and adding the removable nodes to the list as we go, continuing to descend into those removed nodes as we iterate.

thoughts?